### ISSUE TITLE
Route with wildcard and different method can not be matched

### ISSUE THREAD

**Issue by Zombaya** [symfony/symfony#29553]
**Symfony version(s) affected**: 4.1.8 - 4.1.9

**Description**  
A route with a wildcard requested using the OPTIONS-method will not be matched.  However, when viewing the exception using the symfony-profiler will show the correctly matched route.
Up until version 4.1.7 the route would be correctly matched.

This issue looks very similar to issue #29363 to me.

**How to reproduce**
_Route definition_
Note: I prefixed this route with `/dav`.
```xml
<route id="secotrust_sabre_dav" path="/{url}"
           methods="HEAD,GET,POST,OPTIONS,PROPFIND,PROPPATCH,MKCOL,COPY,MOVE,DELETE,LOCK,UNLOCK,PUT,PATCH,REPORT">
   <default key="_controller">secotrust.sabredav.controller::execAction</default>
   <default key="url"></default>
   <requirement key="url">/?.*</requirement>
   <option key="expose">true</option>
</route>
```
_Request_
```
OPTIONS /dav/files/IMIaj8lMzZsnx6mN5gheD7upxd5TqtjlUIMnrud3veqsHQwWZkKSE7pe8iWB4zTJ/ HTTP/1.1
```

**Additional context**  
_Stacktrace_
```
Symfony\Component\Routing\Exception\ResourceNotFoundException:
  at var/cache/dev/srcDevDebugProjectContainerUrlMatcher.php:50
  at srcDevDebugProjectContainerUrlMatcher->match('/dav/files/IMIaj8lMzZsnx6mN5gheD7upxd5TqtjlUIMnrud3veqsHQwWZkKSE7pe8iWB4zTJ/')
     (vendor/symfony/routing/Matcher/UrlMatcher.php:107)
  at Symfony\Component\Routing\Matcher\UrlMatcher->matchRequest(object(Request))
     (vendor/symfony/routing/Router.php:262)
```
_Screenshot of matching route_
![image](https://user-images.githubusercontent.com/13340313/49748239-70f97480-fca5-11e8-96e2-e75567d7d858.png)

_Result with GET-request_
A GET-request would be redirected to the url without the trailing slash.  Up until 4.1.7, this would just give the result from the controller-action.

**Comment by Zombaya** [comment]
I will test this tomorrow morning and get back to you.

**Comment by nicolas-grekas** [comment]
Hum actually it's unrelated - I'll have a look.
Note that the requirement is quite strange, isn't it?
Why don't you use just `.*`?

**Comment by Zombaya** [comment]
It's from an existing bundle so I don't know why the original author opted for that regex.  However `.*` would also not work.  I've written a unittest that works until 4.1.7 but does not work for 4.1.9 and 4.2.1.

```php
// Tests/Matcher/RedirectableUrlMatcherTest.php
public function testAllowedTrailingSlashByRegex()
{
    $coll = new RouteCollection();
    $coll->add('foo', new Route('/{path}',array(),array('path'=>'.*')));

    $context = new RequestContext();
    $context->setMethod('OPTIONS');
    $matcher = $this->getUrlMatcher($coll, $context);
    $this->assertEquals(
        array('path' => 'bar/', '_route' => 'foo'),
        $matcher->match('/bar/')
    );
}
```

**Comment by nicolas-grekas** [comment]
Fixed in #29626
